# CLEANING - PROYECTO SQL

## TO DO :

**PROCESO DE LIMPIEZA:**

a.	Limpiamos la tabla de **Actor**, que hace referencia al archivo *‘actor.csv’*

b.	Limpiamos la tabla de **Category**, que hace referencia al archivo *‘category.csv’*

c.	Limpiamos la tabla de **Film**, que hace referencia al archivo *‘film.csv’*

d.	Limpiamos la tabla de **Inventory**, que hace referencia al archivo *‘inventory.csv’*

e.	Limpiamos la tabla de **Language**, que hace referencia al archivo *‘language.csv’*

f.	Limpiamos la tabla de **old_HDD** que hace referencia al archivo *‘old_HDD.csv’*

g.	Limpiamos la tabla de **Rental** que hace referencia al archivo *‘rental.csv’*

#### Importamos las librerías que puede nos sean necesarias para empezar con la limpieza.

In [1]:
import pandas as pd
import numpy as np
import statistics as stats
import pandas as pd
import numpy as np
import statistics as stats

import warnings
warnings.filterwarnings('ignore')

import pylab as plt
import seaborn as sns

#### Importamos ya los archivos .csv para tenerlos todos en el Jupyter 

In [2]:
actor = pd.read_csv('../data/actor.csv', encoding = 'latin1')
cat = pd.read_csv('../data/category.csv', encoding = 'latin1')
film_df = pd.read_csv('../data/film.csv', encoding = 'latin1')
inv = pd.read_csv('../data/inventory.csv', encoding = 'latin1')
lang = pd.read_csv('../data/language.csv', encoding = 'latin1')
old = pd.read_csv('../data/old_HDD.csv', encoding = 'latin1')
rent = pd.read_csv('../data/rental.csv', encoding = 'latin1')

# Cleaning ACTOR

Importamos el dataframe del csv actor.csv

Observamos si tiene: 
 - Duplicados
 - Nulos

In [3]:
actor.head()

actor_id first_name     last_name          last_update
0         1   PENELOPE       GUINESS  2006-02-15 04:34:33
1         2       NICK      WAHLBERG  2006-02-15 04:34:33
2         3         ED         CHASE  2006-02-15 04:34:33
3         4   JENNIFER         DAVIS  2006-02-15 04:34:33
4         5     JOHNNY  LOLLOBRIGIDA  2006-02-15 04:34:33

In [4]:
actor.duplicated().any()

False

In [5]:
actor.columns

Index(['actor_id', 'first_name', 'last_name', 'last_update'], dtype='object')

In [6]:
actor = actor.rename(columns = {'actor_id': 'Actor_id', 'first_name': 'Name', 'last_name': 'Last_Name', 'last_update': 'Updated'})

In [7]:
actor.columns

Index(['Actor_id', 'Name', 'Last_Name', 'Updated'], dtype='object')

In [8]:
actor.head()

Actor_id      Name     Last_Name              Updated
0         1  PENELOPE       GUINESS  2006-02-15 04:34:33
1         2      NICK      WAHLBERG  2006-02-15 04:34:33
2         3        ED         CHASE  2006-02-15 04:34:33
3         4  JENNIFER         DAVIS  2006-02-15 04:34:33
4         5    JOHNNY  LOLLOBRIGIDA  2006-02-15 04:34:33

In [9]:
actor['Name'].isnull().any()

False

In [10]:
actor['Last_Name'].isnull().any()

False

In [11]:
len(actor.Name)

200

In [12]:
len(actor.Last_Name)

200

In [13]:
actor['Name_complete'] = actor['Name'] + ' ' + actor['Last_Name']

In [14]:
actor['Name_complete'].head()

0       PENELOPE GUINESS
1          NICK WAHLBERG
2               ED CHASE
3         JENNIFER DAVIS
4    JOHNNY LOLLOBRIGIDA
Name: Name_complete, dtype: object

In [15]:
actor.head()

Actor_id      Name     Last_Name              Updated        Name_complete
0         1  PENELOPE       GUINESS  2006-02-15 04:34:33     PENELOPE GUINESS
1         2      NICK      WAHLBERG  2006-02-15 04:34:33        NICK WAHLBERG
2         3        ED         CHASE  2006-02-15 04:34:33             ED CHASE
3         4  JENNIFER         DAVIS  2006-02-15 04:34:33       JENNIFER DAVIS
4         5    JOHNNY  LOLLOBRIGIDA  2006-02-15 04:34:33  JOHNNY LOLLOBRIGIDA

In [16]:
actor.isnull().any()

Actor_id         False
Name             False
Last_Name        False
Updated          False
Name_complete    False
dtype: bool

In [17]:
len(actor.Name_complete)

200

- Ahora que he creado una nueva columna con el nombre completo de los actores, voy a eliminar las columnas de 'Name' y 'Last_Name'

In [18]:
actor = actor.drop(['Name', 'Last_Name'], axis=1)

In [19]:
actor.head()

Actor_id              Updated        Name_complete
0         1  2006-02-15 04:34:33     PENELOPE GUINESS
1         2  2006-02-15 04:34:33        NICK WAHLBERG
2         3  2006-02-15 04:34:33             ED CHASE
3         4  2006-02-15 04:34:33       JENNIFER DAVIS
4         5  2006-02-15 04:34:33  JOHNNY LOLLOBRIGIDA

In [20]:
actor['Name_complete'].value_counts()

Name_complete
SUSAN DAVIS             2
EWAN GOODING            1
DARYL CRAWFORD          1
GRETA KEITEL            1
JANE JACKMAN            1
                       ..
MICHELLE MCCONAUGHEY    1
ADAM GRANT              1
SEAN WILLIAMS           1
GARY PENN               1
THORA TEMPLE            1
Name: count, Length: 199, dtype: int64

In [21]:

# vemos que SUSAN DAVIS está duplicado

# vamos a eliminar este valor duplicado para que no nos de problemas posteriormente. 

# veo que puedo eliminar la fila donde está el Nombre duplicado, ya que no está el 'Autor_id' en ningún otro DF. 


In [22]:
actor_dup = actor[actor.duplicated(subset=['Name_complete'], keep='first')]
actor_dup

Actor_id              Updated Name_complete
109       110  2006-02-15 04:34:33   SUSAN DAVIS

In [23]:
actor = actor.drop_duplicates(subset=['Name_complete'], keep='first')

In [24]:

# una vez que elimino la fila duplicada ya solamente quedan 199


In [25]:
actor_dup = actor[actor.duplicated(subset=['Name_complete'], keep='first')]
actor_dup

Empty DataFrame
Columns: [Actor_id, Updated, Name_complete]
Index: []

In [26]:
actor['Name_complete'].duplicated().value_counts()

Name_complete
False    199
Name: count, dtype: int64

In [27]:
actor['Actor_id']=range(1,len(actor)+1)
actor.reset_index(drop=True, inplace=True)

In [28]:
#pd.set_option('display.max_rows', None)

In [128]:
actor.tail()

Actor_id  Name_complete        Date
194       195    BELA WALKEN  2006-02-15
195       196     REESE WEST  2006-02-15
196       197    MARY KEITEL  2006-02-15
197       198  JULIA FAWCETT  2006-02-15
198       199   THORA TEMPLE  2006-02-15

- La columna de Updated no la voy a eliminar, sino que voy a convertirla en una columna, unicamente con la fecha.

In [30]:
type('Updated')

str

In [31]:
actor['Updated'] = pd.to_datetime(actor['Updated'])

In [32]:
actor['Date'] = actor['Updated'].dt.date
actor['Time'] = actor['Updated'].dt.time

In [33]:
actor = actor.drop(['Updated', 'Time'], axis=1)

In [34]:
actor.head()

Actor_id        Name_complete        Date
0         1     PENELOPE GUINESS  2006-02-15
1         2        NICK WAHLBERG  2006-02-15
2         3             ED CHASE  2006-02-15
3         4       JENNIFER DAVIS  2006-02-15
4         5  JOHNNY LOLLOBRIGIDA  2006-02-15

# Cleaning CATEGORY

Importamos el dataframe del csv category.csv

Observamos si tiene: 
 - Duplicados
 - Nulos

In [35]:
cat

category_id         name          last_update
0             1       Action  2006-02-15 04:46:27
1             2    Animation  2006-02-15 04:46:27
2             3     Children  2006-02-15 04:46:27
3             4     Classics  2006-02-15 04:46:27
4             5       Comedy  2006-02-15 04:46:27
5             6  Documentary  2006-02-15 04:46:27
6             7        Drama  2006-02-15 04:46:27
7             8       Family  2006-02-15 04:46:27
8             9      Foreign  2006-02-15 04:46:27
9            10        Games  2006-02-15 04:46:27
10           11       Horror  2006-02-15 04:46:27
11           12        Music  2006-02-15 04:46:27
12           13          New  2006-02-15 04:46:27
13           14       Sci-Fi  2006-02-15 04:46:27
14           15       Sports  2006-02-15 04:46:27
15           16       Travel  2006-02-15 04:46:27

In [36]:
cat.isnull().any()

category_id    False
name           False
last_update    False
dtype: bool

In [37]:
cat.duplicated().any()

False

In [38]:
cat = cat.drop(['last_update'], axis=1)

In [39]:
cat.head()

category_id       name
0            1     Action
1            2  Animation
2            3   Children
3            4   Classics
4            5     Comedy

In [40]:
cat = cat.rename(columns = {'name': 'Category'})

In [41]:
cat.head()

category_id   Category
0            1     Action
1            2  Animation
2            3   Children
3            4   Classics
4            5     Comedy

# Cleaning FILM

Importamos el dataframe del csv film.csv

Observamos si tiene: 
 - Duplicados
 - Nulos

In [42]:
film_df.head()

film_id             title  \
0        1  ACADEMY DINOSAUR   
1        2    ACE GOLDFINGER   
2        3  ADAPTATION HOLES   
3        4  AFFAIR PREJUDICE   
4        5       AFRICAN EGG   

                                         description  release_year  \
0  A Epic Drama of a Feminist And a Mad Scientist...          2006   
1  A Astounding Epistle of a Database Administrat...          2006   
2  A Astounding Reflection of a Lumberjack And a ...          2006   
3  A Fanciful Documentary of a Frisbee And a Lumb...          2006   
4  A Fast-Paced Documentary of a Pastry Chef And ...          2006   

   language_id  original_language_id  rental_duration  rental_rate  length  \
0            1                   NaN                6         0.99      86   
1            1                   NaN                3         4.99      48   
2            1                   NaN                7         2.99      50   
3            1                   NaN                5         2.99     117   
4            1                   NaN                6         2.99     130   

   replacement_cost rating                  special_features  \
0             20.99     PG  Deleted Scenes,Behind the Scenes   
1             12.99      G           Trailers,Deleted Scenes   
2             18.99  NC-17           Trailers,Deleted Scenes   
3             26.99      G    Commentaries,Behind the Scenes   
4             22.99      G                    Deleted Scenes   

           last_update  
0  2006-02-15 05:03:42  
1  2006-02-15 05:03:42  
2  2006-02-15 05:03:42  
3  2006-02-15 05:03:42  
4  2006-02-15 05:03:42

In [43]:
film_df.shape

(1000, 13)

In [44]:
film_df.duplicated().any()

False

In [45]:
film_df['language_id'].nunique() == 1

True

In [46]:
film_df['release_year'].nunique() == 1

True

In [47]:
film_df.isnull().any()

film_id                 False
title                   False
description             False
release_year            False
language_id             False
original_language_id     True
rental_duration         False
rental_rate             False
length                  False
replacement_cost        False
rating                  False
special_features        False
last_update             False
dtype: bool

In [48]:
film_df.columns

Index(['film_id', 'title', 'description', 'release_year', 'language_id',
       'original_language_id', 'rental_duration', 'rental_rate', 'length',
       'replacement_cost', 'rating', 'special_features', 'last_update'],
      dtype='object')

In [49]:
film_df['original_language_id'].isnull().all()

True

### Eliminamos nulos y repetidos

-  Elimino aquellas columnas que tienen valores iguales y repetidos. 

-  Además, la columna 'original_language_id' tiene todos los valores nulos, por lo que también la eliminamos.

In [50]:
film_df = film_df.drop(['release_year', 'original_language_id', 'last_update', 'rental_rate'], axis=1)


In [51]:
film_df.columns

Index(['film_id', 'title', 'description', 'language_id', 'rental_duration',
       'length', 'replacement_cost', 'rating', 'special_features'],
      dtype='object')

In [52]:
film_df.head()

film_id             title  \
0        1  ACADEMY DINOSAUR   
1        2    ACE GOLDFINGER   
2        3  ADAPTATION HOLES   
3        4  AFFAIR PREJUDICE   
4        5       AFRICAN EGG   

                                         description  language_id  \
0  A Epic Drama of a Feminist And a Mad Scientist...            1   
1  A Astounding Epistle of a Database Administrat...            1   
2  A Astounding Reflection of a Lumberjack And a ...            1   
3  A Fanciful Documentary of a Frisbee And a Lumb...            1   
4  A Fast-Paced Documentary of a Pastry Chef And ...            1   

   rental_duration  length  replacement_cost rating  \
0                6      86             20.99     PG   
1                3      48             12.99      G   
2                7      50             18.99  NC-17   
3                5     117             26.99      G   
4                6     130             22.99      G   

                   special_features  
0  Deleted Scenes,Behind the Scenes  
1           Trailers,Deleted Scenes  
2           Trailers,Deleted Scenes  
3    Commentaries,Behind the Scenes  
4                    Deleted Scenes

In [53]:
film_df['special_features'].nunique()

15

### Limpiamos columnas

- Como el titulo de las columnas está en mayúsculas, por si alguno de las letras no es mayúscula, voy a hacer una comprobación y un cambio de todo a MAYÚSCULAS.

In [54]:
if not film_df['title'].str.isupper().all():
    
    film_df['title'] = film_df['title'].str.upper()

In [55]:
if not film_df['special_features'].str.islower().all():
    
    film_df['special_features'] = film_df['special_features'].str.lower()

In [56]:
film_df.head()

film_id             title  \
0        1  ACADEMY DINOSAUR   
1        2    ACE GOLDFINGER   
2        3  ADAPTATION HOLES   
3        4  AFFAIR PREJUDICE   
4        5       AFRICAN EGG   

                                         description  language_id  \
0  A Epic Drama of a Feminist And a Mad Scientist...            1   
1  A Astounding Epistle of a Database Administrat...            1   
2  A Astounding Reflection of a Lumberjack And a ...            1   
3  A Fanciful Documentary of a Frisbee And a Lumb...            1   
4  A Fast-Paced Documentary of a Pastry Chef And ...            1   

   rental_duration  length  replacement_cost rating  \
0                6      86             20.99     PG   
1                3      48             12.99      G   
2                7      50             18.99  NC-17   
3                5     117             26.99      G   
4                6     130             22.99      G   

                   special_features  
0  deleted scenes,behind the scenes  
1           trailers,deleted scenes  
2           trailers,deleted scenes  
3    commentaries,behind the scenes  
4                    deleted scenes

In [57]:
film_df['rating'].value_counts()

rating
PG-13    223
NC-17    210
R        195
PG       194
G        178
Name: count, dtype: int64

- Para que sea más sencillo, he adaptado un orden del 1 al 5 en los 'rating' de las películas. 

In [58]:
rating_dict = {'PG-13': 1, 'NC-17': 2, 'R': 3, 'PG': 4, 'G': 5}

film_df['rating'] = film_df['rating'].replace(rating_dict)

In [59]:
film_df.head()

film_id             title  \
0        1  ACADEMY DINOSAUR   
1        2    ACE GOLDFINGER   
2        3  ADAPTATION HOLES   
3        4  AFFAIR PREJUDICE   
4        5       AFRICAN EGG   

                                         description  language_id  \
0  A Epic Drama of a Feminist And a Mad Scientist...            1   
1  A Astounding Epistle of a Database Administrat...            1   
2  A Astounding Reflection of a Lumberjack And a ...            1   
3  A Fanciful Documentary of a Frisbee And a Lumb...            1   
4  A Fast-Paced Documentary of a Pastry Chef And ...            1   

   rental_duration  length  replacement_cost  rating  \
0                6      86             20.99       4   
1                3      48             12.99       5   
2                7      50             18.99       2   
3                5     117             26.99       5   
4                6     130             22.99       5   

                   special_features  
0  deleted scenes,behind the scenes  
1           trailers,deleted scenes  
2           trailers,deleted scenes  
3    commentaries,behind the scenes  
4                    deleted scenes

In [60]:
type('replacement_cost')

str

- Cambio el 'replacement_cost' a numeros enteros, ya que el valor anterior ensucia la tabla final. 
- Para ello se pasa el valor a entero mientras redondeamos el número dado al número entero más cercano. 

In [61]:
film_df['replacement_cost'] = round(film_df['replacement_cost']).astype(int)

In [62]:
film_df.head()

film_id             title  \
0        1  ACADEMY DINOSAUR   
1        2    ACE GOLDFINGER   
2        3  ADAPTATION HOLES   
3        4  AFFAIR PREJUDICE   
4        5       AFRICAN EGG   

                                         description  language_id  \
0  A Epic Drama of a Feminist And a Mad Scientist...            1   
1  A Astounding Epistle of a Database Administrat...            1   
2  A Astounding Reflection of a Lumberjack And a ...            1   
3  A Fanciful Documentary of a Frisbee And a Lumb...            1   
4  A Fast-Paced Documentary of a Pastry Chef And ...            1   

   rental_duration  length  replacement_cost  rating  \
0                6      86                21       4   
1                3      48                13       5   
2                7      50                19       2   
3                5     117                27       5   
4                6     130                23       5   

                   special_features  
0  deleted scenes,behind the scenes  
1           trailers,deleted scenes  
2           trailers,deleted scenes  
3    commentaries,behind the scenes  
4                    deleted scenes

# Cleaning INVENTORY

Importamos el dataframe del csv inventory.csv

Observamos si tiene: 
 - Duplicados
 - Nulos

- Elimino aquellas columnas que tienen valores iguales y repetidos.

In [63]:
inv.duplicated().any()

False

In [64]:
inv['film_id'].count()

1000

In [65]:
inv.tail()

inventory_id  film_id  store_id          last_update
995           996      222         2  2006-02-15 05:09:17
996           997      222         2  2006-02-15 05:09:17
997           998      222         2  2006-02-15 05:09:17
998           999      223         2  2006-02-15 05:09:17
999          1000      223         2  2006-02-15 05:09:17

In [66]:
inv = inv.drop(['last_update'], axis=1)

In [67]:
inv.head()

inventory_id  film_id  store_id
0             1        1         1
1             2        1         1
2             3        1         1
3             4        1         1
4             5        1         2

- Observamos si el número de valores únicos de las películas que están en el inventario (tabla 'INVENTORY') coinciden con las películas de la tabla 'FILM_DF'. 
- Podemos hacer esto porque tenemos un valor en una columna que coincide, el 'film_id'. 

In [68]:
inv.shape

(1000, 3)

In [69]:
len(inv['film_id'].unique())

207

In [70]:
len(film_df['film_id'].unique ())

1000

In [71]:
# Unir los dataframes utilizando la función "merge"
merged_df = pd.merge(inv, film_df, on="film_id")

# Obtener la lista de películas que coinciden en ambas tablas
common_movies = merged_df["film_id"].unique()
common_movies

array([  1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,  13,
        15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,  26,  27,
        28,  29,  30,  31,  32,  34,  35,  37,  39,  40,  42,  43,  44,
        45,  46,  47,  48,  49,  50,  51,  52,  53,  54,  55,  56,  57,
        58,  59,  60,  61,  62,  63,  64,  65,  66,  67,  68,  69,  70,
        71,  72,  73,  74,  75,  76,  77,  78,  79,  80,  81,  82,  83,
        84,  85,  86,  88,  89,  90,  91,  92,  93,  94,  95,  96,  97,
        98,  99, 100, 101, 102, 103, 104, 105, 106, 107, 109, 110, 111,
       112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124,
       125, 126, 127, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138,
       139, 140, 141, 142, 143, 145, 146, 147, 149, 150, 151, 152, 153,
       154, 155, 156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166,
       167, 168, 169, 170, 172, 173, 174, 175, 176, 177, 178, 179, 180,
       181, 182, 183, 184, 185, 186, 187, 188, 189, 190, 191, 19

- Al hacer esto vemos que en efecto hay 1000 películas en la tabla 'INV', y 1000 películas en la tabla 'FILM_DF'. Sin embargo, los valores únicos de cada una no coindicen ya que el inventario muestra únicamente 223 valores únicos (una misma película está en varias tiendas).


- Además, las películas de la tabla 'INV' coinciden con las 223 primeras películas de la tabla 'FILM_DF'.


# Cleaning LANGUAGE

Importamos el dataframe del csv language.csv

Observamos si tiene: 
 - Duplicados
 - Nulos
 
Renombramos la columna 'name' para que sea más visual y facil de utilizar posteriormente

In [72]:
lang

language_id      name          last_update
0            1   English  2006-02-15 05:02:19
1            2   Italian  2006-02-15 05:02:19
2            3  Japanese  2006-02-15 05:02:19
3            4  Mandarin  2006-02-15 05:02:19
4            5    French  2006-02-15 05:02:19
5            6    German  2006-02-15 05:02:19

In [73]:
lang = lang.drop(['last_update'], axis=1)

In [74]:
lang = lang.rename(columns = {'name': 'Language'})

In [75]:
lang

language_id  Language
0            1   English
1            2   Italian
2            3  Japanese
3            4  Mandarin
4            5    French
5            6    German

# Cleaning old_HDD


Observamos si tiene: 
 - Duplicados
 - Nulos

In [76]:
old.head()

first_name last_name                  title  release_year  category_id
0   PENELOPE   GUINESS       ACADEMY DINOSAUR          2006            6
1   PENELOPE   GUINESS   ANACONDA CONFESSIONS          2006            2
2   PENELOPE   GUINESS            ANGELS LIFE          2006           13
3   PENELOPE   GUINESS  BULWORTH COMMANDMENTS          2006           10
4   PENELOPE   GUINESS          CHEAPER CLYDE          2006           14

In [77]:
old.duplicated().any()

False

In [78]:
old.isnull().any()

first_name      False
last_name       False
title           False
release_year    False
category_id     False
dtype: bool

In [79]:
old.count()

first_name      1000
last_name       1000
title           1000
release_year    1000
category_id     1000
dtype: int64

In [80]:
old['release_year'].nunique() == 1

True

- Los valores de la columna 'release_year' son únicos, por lo que voy a eliminar esta columna. 

In [81]:
old = old.drop(['release_year'], axis=1)

In [82]:
old.columns

Index(['first_name', 'last_name', 'title', 'category_id'], dtype='object')

In [83]:
old = old.rename(columns = {'first_name': 'Name', 'last_name': 'Last_Name'})

In [84]:
old.head()

Name Last_Name                  title  category_id
0  PENELOPE   GUINESS       ACADEMY DINOSAUR            6
1  PENELOPE   GUINESS   ANACONDA CONFESSIONS            2
2  PENELOPE   GUINESS            ANGELS LIFE           13
3  PENELOPE   GUINESS  BULWORTH COMMANDMENTS           10
4  PENELOPE   GUINESS          CHEAPER CLYDE           14

In [85]:
old['Name_complete'] = old['Name'] + ' ' + old['Last_Name']

In [86]:
old.head()

Name Last_Name                  title  category_id     Name_complete
0  PENELOPE   GUINESS       ACADEMY DINOSAUR            6  PENELOPE GUINESS
1  PENELOPE   GUINESS   ANACONDA CONFESSIONS            2  PENELOPE GUINESS
2  PENELOPE   GUINESS            ANGELS LIFE           13  PENELOPE GUINESS
3  PENELOPE   GUINESS  BULWORTH COMMANDMENTS           10  PENELOPE GUINESS
4  PENELOPE   GUINESS          CHEAPER CLYDE           14  PENELOPE GUINESS

In [87]:
old = old.drop(['Name', 'Last_Name'], axis=1)

In [88]:
old.head()

title  category_id     Name_complete
0       ACADEMY DINOSAUR            6  PENELOPE GUINESS
1   ANACONDA CONFESSIONS            2  PENELOPE GUINESS
2            ANGELS LIFE           13  PENELOPE GUINESS
3  BULWORTH COMMANDMENTS           10  PENELOPE GUINESS
4          CHEAPER CLYDE           14  PENELOPE GUINESS

In [89]:
# hay duplicados en los nombres de los títulos porque más de un 'actor', está en un mismo 'Title'

len(old['Name_complete'].unique())

39

In [90]:
len(old['title'].unique()) 

614

In [91]:
len(film_df['title'].unique())

1000

- Le voy a poner un índice al dataframe 'old'. 
- Esto me facilitará su uso posterior en SQL 

In [92]:
old['old_id'] = range(1, len(old)+1)

In [93]:
old.head()

title  category_id     Name_complete  old_id
0       ACADEMY DINOSAUR            6  PENELOPE GUINESS       1
1   ANACONDA CONFESSIONS            2  PENELOPE GUINESS       2
2            ANGELS LIFE           13  PENELOPE GUINESS       3
3  BULWORTH COMMANDMENTS           10  PENELOPE GUINESS       4
4          CHEAPER CLYDE           14  PENELOPE GUINESS       5

In [94]:
column_order = ['old_id', 'Title', 'Name_complete', 'category_id']

In [95]:
old = old.reindex(columns=column_order)

In [96]:
old.head()

old_id  Title     Name_complete  category_id
0       1    NaN  PENELOPE GUINESS            6
1       2    NaN  PENELOPE GUINESS            2
2       3    NaN  PENELOPE GUINESS           13
3       4    NaN  PENELOPE GUINESS           10
4       5    NaN  PENELOPE GUINESS           14

In [97]:
old = old.set_index('old_id')

- Añado la columna 'Actor_id' en la tabla 'old'. Esto puedo hacerlo ya que tengo un elemento común entre ambas tablas, el 'Name_complete' está en ambas columnas. 

In [98]:
# Seleccionar las columnas que deseas mantener en la tabla resultante y renombrarlas según sea necesario

In [99]:
old = old.merge(actor[['Name_complete', 'Actor_id']], on = 'Name_complete', how = 'left') 
old.head()

Title     Name_complete  category_id  Actor_id
0    NaN  PENELOPE GUINESS            6         1
1    NaN  PENELOPE GUINESS            2         1
2    NaN  PENELOPE GUINESS           13         1
3    NaN  PENELOPE GUINESS           10         1
4    NaN  PENELOPE GUINESS           14         1

# Cleaning RENTAL

Observamos si tiene: 
 - Duplicados
 - Nulos

In [100]:
rent.head()

rental_id          rental_date  inventory_id  customer_id  \
0          1  2005-05-24 22:53:30           367          130   
1          2  2005-05-24 22:54:33          1525          459   
2          3  2005-05-24 23:03:39          1711          408   
3          4  2005-05-24 23:04:41          2452          333   
4          5  2005-05-24 23:05:21          2079          222   

           return_date  staff_id          last_update  
0  2005-05-26 22:04:30         1  2006-02-15 21:30:53  
1  2005-05-28 19:40:33         1  2006-02-15 21:30:53  
2  2005-06-01 22:12:39         1  2006-02-15 21:30:53  
3  2005-06-03 01:43:41         2  2006-02-15 21:30:53  
4  2005-06-02 04:33:21         1  2006-02-15 21:30:53

In [101]:
rent.duplicated().any()

False

In [102]:
rent.isnull().any()

rental_id       False
rental_date     False
inventory_id    False
customer_id     False
return_date     False
staff_id        False
last_update     False
dtype: bool

- No hay nulos
- No hay filas duplicadas.

In [103]:
rent = rent.drop(['last_update'], axis=1)

In [104]:
rent.head()

rental_id          rental_date  inventory_id  customer_id  \
0          1  2005-05-24 22:53:30           367          130   
1          2  2005-05-24 22:54:33          1525          459   
2          3  2005-05-24 23:03:39          1711          408   
3          4  2005-05-24 23:04:41          2452          333   
4          5  2005-05-24 23:05:21          2079          222   

           return_date  staff_id  
0  2005-05-26 22:04:30         1  
1  2005-05-28 19:40:33         1  
2  2005-06-01 22:12:39         1  
3  2005-06-03 01:43:41         2  
4  2005-06-02 04:33:21         1

In [105]:
column_order = ['rental_id', 'inventory_id', 'customer_id', 'staff_id', 'rental_date', 'return_date']

In [106]:
rent = rent.reindex(columns=column_order)

In [107]:
rent.head()

rental_id  inventory_id  customer_id  staff_id          rental_date  \
0          1           367          130         1  2005-05-24 22:53:30   
1          2          1525          459         1  2005-05-24 22:54:33   
2          3          1711          408         1  2005-05-24 23:03:39   
3          4          2452          333         2  2005-05-24 23:04:41   
4          5          2079          222         1  2005-05-24 23:05:21   

           return_date  
0  2005-05-26 22:04:30  
1  2005-05-28 19:40:33  
2  2005-06-01 22:12:39  
3  2005-06-03 01:43:41  
4  2005-06-02 04:33:21

In [108]:
rent.shape

(1000, 6)

In [109]:
rent.rental_id.value_counts().head()

rental_id
1      1
673    1
660    1
661    1
662    1
Name: count, dtype: int64

In [110]:
len(rent['inventory_id'].unique()) 

1000

In [111]:
len(inv['inventory_id'].unique())

1000

In [112]:
# Unir los dataframes utilizando la función "merge"
merged_df = pd.merge(rent, inv, on = 'inventory_id')

# Obtener la lista de inventory que coinciden en ambas tablas
common_inv_id = merged_df['inventory_id'].unique()
common_inv_id

array([367, 389, 830, 146, 727, 611, 403, 380, 330, 261,  79, 239, 517,
       466, 470, 139, 775, 178, 535, 617, 373, 794, 749, 872,  37, 212,
       952, 815, 399, 103, 327, 655, 811, 847, 633, 909, 288, 887, 865,
       851, 984, 506, 190, 776, 674, 303, 102, 729, 764, 679, 651, 214,
       511, 715, 241, 503, 131, 368,  30, 435, 207, 971, 656, 182, 661,
       350, 314, 485, 249, 187, 593,  26, 751, 946, 141,   6, 833, 360,
       944, 955, 910, 434, 691, 263, 145, 752, 553, 258, 461, 619, 129,
       354, 723, 205, 597, 826,  20, 339, 802, 424,  72, 393, 184, 458,
       601,  71, 667, 225, 234, 266, 992, 530, 487,  23, 158, 644, 928,
        53, 375, 898,  95, 473, 377, 638, 595, 821, 599, 270, 280, 989,
       452, 425,  58, 495, 140, 739, 126, 331, 848,  87, 797, 227, 697,
       907, 742, 800, 635, 592, 130, 500, 372, 819, 194,  85, 267, 933,
       242, 769, 600, 273, 581, 730, 734, 515, 687, 353, 245, 451, 533,
        88,  39, 763, 708, 379, 609, 441, 785, 630, 186, 623, 71

In [113]:
len(common_inv_id)

212

- Observamos si el número de valores únicos del 'inventory_id' que están en el inventario (tabla 'INVENTORY') coinciden con el 'inventory_id' de la tabla 'RENT'.

- Podemos hacer esto porque tenemos un valor en una columna que coincide, el 'inventory_id'. 

- Al hacer esto vemos que en efecto hay 1000 'inventory_id' en la tabla 'INV', y 1000 'inventory_id' en la tabla 'RENT', todos ellos valores únicos. Sin embargo, los valores únicos de cada una no coindicen ya que el inventario común muestra únicamente 212 valores únicos.

**CONCLUSIÓN INVENTORY**

- Como conclusión en este caso, puede que aunque tengamos un inventory con películas que no se alquilan, y **películas que se alquilan que no están en el inventario**. 

- Tampoco coincidían el **número de 'FILM_ID' de la tabla 'INVENTORY' y la tabla 'FILM_DF'.**

- Esto nos hace pensar que el **inventario no está completo**.

### Tiempo de alquiler

- Voy a limpiar las columnas de 'rental_date' y 'return_date' con el fin de crear una única columna que me informe cuántos días tarda una persona en devolver una película alquilada. 

In [114]:
type('rental_date')

str

In [115]:
rent['rental_date'] = pd.to_datetime(rent['rental_date'])

In [116]:
rent['Rent_Date'] = rent['rental_date'].dt.date

In [117]:
rent['return_date'] = pd.to_datetime(rent['return_date'])

In [118]:
rent['Return_Date'] = rent['return_date'].dt.date

In [119]:
rent.head()

rental_id  inventory_id  customer_id  staff_id         rental_date  \
0          1           367          130         1 2005-05-24 22:53:30   
1          2          1525          459         1 2005-05-24 22:54:33   
2          3          1711          408         1 2005-05-24 23:03:39   
3          4          2452          333         2 2005-05-24 23:04:41   
4          5          2079          222         1 2005-05-24 23:05:21   

          return_date   Rent_Date Return_Date  
0 2005-05-26 22:04:30  2005-05-24  2005-05-26  
1 2005-05-28 19:40:33  2005-05-24  2005-05-28  
2 2005-06-01 22:12:39  2005-05-24  2005-06-01  
3 2005-06-03 01:43:41  2005-05-24  2005-06-03  
4 2005-06-02 04:33:21  2005-05-24  2005-06-02

In [120]:
rent['Days_rent'] = (rent['return_date'] - rent['rental_date']).dt.days

In [121]:
rent.head()

rental_id  inventory_id  customer_id  staff_id         rental_date  \
0          1           367          130         1 2005-05-24 22:53:30   
1          2          1525          459         1 2005-05-24 22:54:33   
2          3          1711          408         1 2005-05-24 23:03:39   
3          4          2452          333         2 2005-05-24 23:04:41   
4          5          2079          222         1 2005-05-24 23:05:21   

          return_date   Rent_Date Return_Date  Days_rent  
0 2005-05-26 22:04:30  2005-05-24  2005-05-26          1  
1 2005-05-28 19:40:33  2005-05-24  2005-05-28          3  
2 2005-06-01 22:12:39  2005-05-24  2005-06-01          7  
3 2005-06-03 01:43:41  2005-05-24  2005-06-03          9  
4 2005-06-02 04:33:21  2005-05-24  2005-06-02          8

In [122]:
# Ya no necesito las columnas anteriores de 'rental_date' y 'return_date'. Las elimino.

- Ahora que ya he hecho la limpieza de completa de la fecha de 'rental_date' y 'return_date', entonces voy a eliminar estas dos, además de las dos que he creado para contar los días que una persona tarda en devolver una película alquilada ('Days_rent').

In [123]:
rent = rent.drop(['rental_date', 'return_date', 'Rent_Date', 'Return_Date' ], axis=1)

In [124]:
rent.head()

rental_id  inventory_id  customer_id  staff_id  Days_rent
0          1           367          130         1          1
1          2          1525          459         1          3
2          3          1711          408         1          7
3          4          2452          333         2          9
4          5          2079          222         1          8

In [125]:
rent['Days_rent'].value_counts()

Days_rent
1    122
8    117
7    111
2    110
5    109
3    107
4    104
6    100
0     63
9     57
Name: count, dtype: int64

rent = rent.merge(film_df[['rental_duration', 'film_id']], on = 'Days_rent', how = 'left')

# CONEXIÓN A SQL Y CREACIÓN DE BBDD

In [127]:
import mysql.connector as conn

Vamos a exportar los archivos .csv limpios para comenzar con el estudio en MySQL

In [129]:
#actor.to_csv('../data_clean/actor_clean.csv', index = False)

In [130]:
#cat.to_csv('../data_clean/cat_clean.csv', index = False)

In [131]:
#film_df.to_csv('../data_clean/film_clean.csv', index = False)

In [132]:
#inv.to_csv('../data_clean/inv_clean.csv', index = False)

In [133]:
#lang.to_csv('../data_clean/lang_clean.csv', index = False)

In [137]:
#old.to_csv('../data_clean/old_HDD_clean.csv', index = False)

In [138]:
#rent.to_csv('../data_clean/rental_clean.csv', index = False)

### Observo ahora las columnas de cada uno de los DataFrame, para ver la relación final de todos estos.

In [ ]:
actor.columns

In [ ]:
lang.columns

In [ ]:
film_df.columns

In [ ]:
inv.columns

In [ ]:
cat.columns

In [ ]:
old.columns

In [ ]:
rent.columns